### ISSUE TITLE
Problem with deduplication of alerts between Grafana Internal alertmanager

### ISSUE THREAD

**Issue by mgronst** [grafana/grafana#55938]
Running two separate instances of Grafana v. 9.1.6 in Docker Containers. Taken the Grafana unified alerting into use, and configured the internal alertmanager high availability mode.

The full alert configuration is retrieved by API from one Grafana instance and provisioned on the other one. However both Grafana instances is firing alerts towards the configured Slack endpoint. 

I can see in the logs that the high availability functionality is synchronized.

`logger=ngalert.multiorg.alertmanager component=cluster t=2022-09-28T14:47:05.412199748Z level=info msg="gossip settled; proceeding" elapsed=10.002850312s`

But also see a lot of warnings regarding alertmanager state:
```
logger=ngalert.multiorg.alertmanager component=cluster t=2022-09-28T14:47:25.000883237Z level=warn received="unknown state key" len=8702 key=notificationlog:2
logger=ngalert.multiorg.alertmanager component=cluster t=2022-09-28T14:47:25.000922612Z level=warn received="unknown state key" len=8702 key=silences:2
```
The other Grafana instance has the same warnings, but a different number for what i believe is the state key:
```
logger=ngalert.multiorg.alertmanager component=cluster t=2022-09-28T14:48:36.956326088Z level=warn received="unknown state key" len=8321 key=notificationlog:3
logger=ngalert.multiorg.alertmanager component=cluster t=2022-09-28T14:48:36.956394037Z level=warn received="unknown state key" len=8321 key=silences:3

```

**Comment by JacobsonMT** [comment]
Hi @mgronst, would it be possible to enable debug logging and give us a larger time range of logs? Also, seeing your configuration related to HA would help give us an idea if it's a possible misconfiguration.

**Comment by mgronst** [comment]
@JacobsonMT Please find attached log files for both Grafana instances configured in HA with debug enabled, from 
a restart to when first duplicate alert is triggered.
Also added the relevant HA alert config from both instances.
I have removed several lines from the debug log that could contain private information about our environment.
[grafana_config_debug.zip](https://github.com/grafana/grafana/files/9754940/grafana_config_debug.zip)

**Comment by mgronst** [comment]
Any new input needed, or any tips on configuration changes we can test?

**Comment by mgronst** [comment]
Upgraded to Grafana v. 9.3.2, and we still see the same issue. No deduplication of any alarms. Any idea of solution on this? @armandgrillet @JacobsonMT @dprokop

**Comment by gotjosh** [comment]
The state key is a combination of the component + the tenant_id - in this case the OrgID that the Grafana instance uses.

Because Grafana is a multi-tenant system by default, we use this combination to ensure that the state from tenant A does not get mixed up with the state of tenant B.

It seems to me that the grafana instances you have configured have two completely separate OrgIDs for the main tenant that you're testing with somehow.

**Comment by mgronst** [comment]
@gotjosh How can I do more troubleshooting on this? From the alerts that are triggered, and received in duplicate, they all seem to have the same OrgID in the source. 

Example from main instance:
`Source: https://graf1.*/alerting/grafana/0qF4MmVVk/view?orgId=1`
`Dashboard: https://graf1.*/d/KdY74xG4k?orgId=1`

Duplicate alert from redundant system:
`Source: https://graf2.*/alerting/grafana/0qF4MmVVk/view?orgId=1`
`Dashboard: https://graf2.*/d/KdY74xG4k?orgId=1`

**Comment by gotjosh** [comment]
Looking at your logs - I can see that you have not successfully connected both Grafana instances, `ha_peers` requires that you list all of the grafana instances you're trying to include as part of the mesh. If you have 3 Grafana instances, you should include 3 IPs or DNS names if you're using k8s.

We expose a high number metrics to aid you with debugging, everything under `alertmanager_cluster` should give you an accurate view of the state of your cluster - what do you see in these metrics?

**Comment by mgronst** [comment]
@gotjosh Added both peers in my cluster to `ha_peers` on both instances, followed by a restart. Still the same issue with duplicate alerts is present.

Where to I find the `alertmanager_cluster` metrics. We are not utilizing any internal metrics, only external Prometheus integrations.

**Comment by gotjosh** [comment]
Please note that we're entering the land of support as oppose to an actual issue with the software.

For those, https://community.grafana.com/ is the best place to ask for such questions. 

`alertmanager_cluster` metrics are exposed by Grafana and need to be collected by your external prometheus - I don't really understand what you mean with Prometheus external integrations.

**Comment by mgronst** [comment]
@gotjosh Thanks for clarifying which metrics you referred to. I believe we have configured everything right according to the documentation available, and that there is some software issue related to the deduplication. If you see it otherwise, please close the issue, and we can follow up in community channels.

Below are the recent values of the `alertmanager_cluster` metrics for both peers. 

**PEER1:**
```
# HELP alertmanager_cluster_alive_messages_total Total number of received alive messages.
# TYPE alertmanager_cluster_alive_messages_total counter
alertmanager_cluster_alive_messages_total{peer="01GQQWFBW53NFM8284RR0YR4QS"} 70215
alertmanager_cluster_alive_messages_total{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX"} 70218
# HELP alertmanager_cluster_failed_peers Number indicating the current number of failed peers in the cluster.
# TYPE alertmanager_cluster_failed_peers gauge
alertmanager_cluster_failed_peers 0
# HELP alertmanager_cluster_health_score Health score of the cluster. Lower values are better and zero means 'totally healthy'.
# TYPE alertmanager_cluster_health_score gauge
alertmanager_cluster_health_score 0
# HELP alertmanager_cluster_members Number indicating current number of members in cluster.
# TYPE alertmanager_cluster_members gauge
alertmanager_cluster_members 2
# HELP alertmanager_cluster_messages_pruned_total Total number of cluster messages pruned.
# TYPE alertmanager_cluster_messages_pruned_total counter
alertmanager_cluster_messages_pruned_total 0
# HELP alertmanager_cluster_messages_queued Number of cluster messages which are queued.
# TYPE alertmanager_cluster_messages_queued gauge
alertmanager_cluster_messages_queued 0
# HELP alertmanager_cluster_messages_received_size_total Total size of cluster messages received.
# TYPE alertmanager_cluster_messages_received_size_total counter
alertmanager_cluster_messages_received_size_total{msg_type="full_state"} 5.66904075e+08
alertmanager_cluster_messages_received_size_total{msg_type="update"} 636351
# HELP alertmanager_cluster_messages_received_total Total number of cluster messages received.
# TYPE alertmanager_cluster_messages_received_total counter
alertmanager_cluster_messages_received_total{msg_type="full_state"} 70098
alertmanager_cluster_messages_received_total{msg_type="update"} 4927
# HELP alertmanager_cluster_messages_sent_size_total Total size of cluster messages sent.
# TYPE alertmanager_cluster_messages_sent_size_total counter
alertmanager_cluster_messages_sent_size_total{msg_type="full_state"} 5.93628309e+08
alertmanager_cluster_messages_sent_size_total{msg_type="update"} 635052
# HELP alertmanager_cluster_messages_sent_total Total number of cluster messages sent.
# TYPE alertmanager_cluster_messages_sent_total counter
alertmanager_cluster_messages_sent_total{msg_type="full_state"} 70103
alertmanager_cluster_messages_sent_total{msg_type="update"} 4917
# HELP alertmanager_cluster_peer_info A metric with a constant '1' value labeled by peer name.
# TYPE alertmanager_cluster_peer_info gauge
alertmanager_cluster_peer_info{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX"} 1
# HELP alertmanager_cluster_peers_joined_total A counter of the number of peers that have joined.
# TYPE alertmanager_cluster_peers_joined_total counter
alertmanager_cluster_peers_joined_total 6
# HELP alertmanager_cluster_peers_left_total A counter of the number of peers that have left.
# TYPE alertmanager_cluster_peers_left_total counter
alertmanager_cluster_peers_left_total 4
# HELP alertmanager_cluster_peers_update_total A counter of the number of peers that have updated metadata.
# TYPE alertmanager_cluster_peers_update_total counter
alertmanager_cluster_peers_update_total 0
# HELP alertmanager_cluster_pings_seconds Histogram of latencies for ping messages.
# TYPE alertmanager_cluster_pings_seconds histogram
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.005"} 1.212469e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.01"} 1.218415e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.025"} 1.219792e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.05"} 1.22003e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.1"} 1.220151e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.25"} 1.22017e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="0.5"} 1.22017e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFBW53NFM8284RR0YR4QS",le="+Inf"} 1.22017e+06
alertmanager_cluster_pings_seconds_sum{peer="01GQQWFBW53NFM8284RR0YR4QS"} 3041.9061693349463
alertmanager_cluster_pings_seconds_count{peer="01GQQWFBW53NFM8284RR0YR4QS"} 1.22017e+06
# HELP alertmanager_cluster_reconnections_failed_total A counter of the number of failed cluster peer reconnection attempts.
# TYPE alertmanager_cluster_reconnections_failed_total counter
alertmanager_cluster_reconnections_failed_total 4
# HELP alertmanager_cluster_reconnections_total A counter of the number of cluster peer reconnections.
# TYPE alertmanager_cluster_reconnections_total counter
alertmanager_cluster_reconnections_total 4
# HELP alertmanager_cluster_refresh_join_failed_total A counter of the number of failed cluster peer joined attempts via refresh.
# TYPE alertmanager_cluster_refresh_join_failed_total counter
alertmanager_cluster_refresh_join_failed_total 2
# HELP alertmanager_cluster_refresh_join_total A counter of the number of cluster peer joined via refresh.
# TYPE alertmanager_cluster_refresh_join_total counter
alertmanager_cluster_refresh_join_total 3
```

**PEER2:**
```
# HELP alertmanager_cluster_alive_messages_total Total number of received alive messages.
# TYPE alertmanager_cluster_alive_messages_total counter
alertmanager_cluster_alive_messages_total{peer="01GQQWFBW53NFM8284RR0YR4QS"} 70216
alertmanager_cluster_alive_messages_total{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX"} 70221
# HELP alertmanager_cluster_failed_peers Number indicating the current number of failed peers in the cluster.
# TYPE alertmanager_cluster_failed_peers gauge
alertmanager_cluster_failed_peers 0
# HELP alertmanager_cluster_health_score Health score of the cluster. Lower values are better and zero means 'totally healthy'.
# TYPE alertmanager_cluster_health_score gauge
alertmanager_cluster_health_score 0
# HELP alertmanager_cluster_members Number indicating current number of members in cluster.
# TYPE alertmanager_cluster_members gauge
alertmanager_cluster_members 2
# HELP alertmanager_cluster_messages_pruned_total Total number of cluster messages pruned.
# TYPE alertmanager_cluster_messages_pruned_total counter
alertmanager_cluster_messages_pruned_total 0
# HELP alertmanager_cluster_messages_queued Number of cluster messages which are queued.
# TYPE alertmanager_cluster_messages_queued gauge
alertmanager_cluster_messages_queued 0
# HELP alertmanager_cluster_messages_received_size_total Total size of cluster messages received.
# TYPE alertmanager_cluster_messages_received_size_total counter
alertmanager_cluster_messages_received_size_total{msg_type="full_state"} 5.93594467e+08
alertmanager_cluster_messages_received_size_total{msg_type="update"} 634922
# HELP alertmanager_cluster_messages_received_total Total number of cluster messages received.
# TYPE alertmanager_cluster_messages_received_total counter
alertmanager_cluster_messages_received_total{msg_type="full_state"} 70096
alertmanager_cluster_messages_received_total{msg_type="update"} 4916
# HELP alertmanager_cluster_messages_sent_size_total Total size of cluster messages sent.
# TYPE alertmanager_cluster_messages_sent_size_total counter
alertmanager_cluster_messages_sent_size_total{msg_type="full_state"} 5.66912149e+08
alertmanager_cluster_messages_sent_size_total{msg_type="update"} 636612
# HELP alertmanager_cluster_messages_sent_total Total number of cluster messages sent.
# TYPE alertmanager_cluster_messages_sent_total counter
alertmanager_cluster_messages_sent_total{msg_type="full_state"} 70101
alertmanager_cluster_messages_sent_total{msg_type="update"} 4929
# HELP alertmanager_cluster_peer_info A metric with a constant '1' value labeled by peer name.
# TYPE alertmanager_cluster_peer_info gauge
alertmanager_cluster_peer_info{peer="01GQQWFBW53NFM8284RR0YR4QS"} 1
# HELP alertmanager_cluster_peers_joined_total A counter of the number of peers that have joined.
# TYPE alertmanager_cluster_peers_joined_total counter
alertmanager_cluster_peers_joined_total 10
# HELP alertmanager_cluster_peers_left_total A counter of the number of peers that have left.
# TYPE alertmanager_cluster_peers_left_total counter
alertmanager_cluster_peers_left_total 8
# HELP alertmanager_cluster_peers_update_total A counter of the number of peers that have updated metadata.
# TYPE alertmanager_cluster_peers_update_total counter
alertmanager_cluster_peers_update_total 0
# HELP alertmanager_cluster_pings_seconds Histogram of latencies for ping messages.
# TYPE alertmanager_cluster_pings_seconds histogram
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.005"} 1.212306e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.01"} 1.218557e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.025"} 1.219982e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.05"} 1.220266e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.1"} 1.220384e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.25"} 1.220413e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="0.5"} 1.220413e+06
alertmanager_cluster_pings_seconds_bucket{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX",le="+Inf"} 1.220413e+06
alertmanager_cluster_pings_seconds_sum{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX"} 3120.2357489408846
alertmanager_cluster_pings_seconds_count{peer="01GQQWFDYJ3T7K9AHYZHJ1YVEX"} 1.220413e+06
# HELP alertmanager_cluster_reconnections_failed_total A counter of the number of failed cluster peer reconnection attempts.
# TYPE alertmanager_cluster_reconnections_failed_total counter
alertmanager_cluster_reconnections_failed_total 2
# HELP alertmanager_cluster_reconnections_total A counter of the number of cluster peer reconnections.
# TYPE alertmanager_cluster_reconnections_total counter
alertmanager_cluster_reconnections_total 5
# HELP alertmanager_cluster_refresh_join_failed_total A counter of the number of failed cluster peer joined attempts via refresh.
# TYPE alertmanager_cluster_refresh_join_failed_total counter
alertmanager_cluster_refresh_join_failed_total 1
# HELP alertmanager_cluster_refresh_join_total A counter of the number of cluster peer joined via refresh.
# TYPE alertmanager_cluster_refresh_join_total counter
alertmanager_cluster_refresh_join_total 5
```

**Comment by mgronst** [comment]
Anyone that has looked into this issue lately?

**Comment by iXingo** [comment]
same here

**Comment by MichaelMayorov** [comment]
In my case it solved simply by restarting entire grafana cluster.

Problem itself seems closely related with connectivity issue. We got 3 grafana servers cluster and sometime got 2 notifications at the same time. After proper inspection I found out that our network equipment blocking alertmanager connection between 2 specific nodes. Permitting should fix issue, but it won't. I've checked that all alertmanagers are able to connect each other, DNS working properly and still got 2 alerts. Only after restarting grafana servers one by one it finally seems working correct.

**Comment by iXingo** [comment]
#76809

**Comment by zumby** [comment]
@mgronst we have been experiencing this issue for quite long. I believe from v9.x.x. First we had our Grafana deployed as HA on VMs - we tried setup the HA - did not resolve duplicated alerts.
Then we switched to Kubernetes and ensured 100% correct setup according to the docs - still getting duplicated alerts **sometimes**.

But i suppose the key thing is this: https://grafana.com/docs/grafana/latest/alerting/set-up/configure-high-availability/#enable-alerting-high-availability-using-redis
> Prevent duplicate notifications
> In high-availability mode, each Grafana instance runs its own pre-configured alertmanager to handle alert notifications.
> 
> When multiple Grafana instances are running, all alert rules are evaluated on each instance. By default, each instance sends firing alerts to its respective alertmanager. This results in notification handling being duplicated across all running Grafana instances.
> 
> Alertmanagers in HA mode communicate with each other to coordinate notification delivery. However, this setup can sometimes lead to duplicated or out-of-order notifications. By design, HA prioritizes sending duplicate notifications over the risk of missing notifications.
> 
> To avoid duplicate notifications, you can configure a shared alertmanager to manage notifications for all Grafana instances. For more information, refer to [add an external alertmanager](https://grafana.com/docs/grafana/latest/alerting/set-up/configure-alertmanager/).

Today we have tried 2 things 
- implement separate Prometheus Alertmanager and pass alerts to it >> resolves duplicated alerts issue
- implement Redis and all the `ha_redis_*` configs >> resolves duplicated alerts issue

The bottom line imo is that Grafana is simply unable to handle deduplication properly and you might be right about some software related issue. You are not alone here :)

Cheers.

**Comment by yuri-tceretian** [comment]
The Alertmanager is AP system (in terms of CAP theory), and therefore the duplicated notifications are expected by design, especially in the case of network partition. However, we acknowledged that sometime in HA mode Grafana sends too many duplicated notifications. Usually, it is related to few factors:

We've improved deduplication in https://github.com/grafana/grafana/pull/99825. In Grafana 11.6.0 this went as experimental change and requires flags `alertingAlertmanagerExtraDedupStage` and `alertingAlertmanagerExtraDedupStageStopPipeline` to be enabled.
In Grafana 12 it will be enabled by default. 
However, although it greatly reduces the duplicated notifications in HA, it does not eliminate them completely. There are still race conditions possible especially is volatile environments such Kubernetes when Grafana instances can be restarted arbitrarily. This breaks alignment of so called "flushing" of the instances, and they can align the way that they send notification at almost the same time that is not enough for them to report about successful notification. 
https://github.com/grafana/prometheus-alertmanager/pull/104 shall reduce the duplicated

**Comment by aifraenkel** [comment]
Closing this issue as the PRs mentioned above were already merged.
